File: src/moltauth/signing.py
"""HTTP Signatures implementation using Ed25519.

Implements RFC 9421 (HTTP Message Signatures) with Ed25519.
"""

import base64
import hashlib
import re
import time
from datetime import datetime, timezone
from typing import Dict, List, Optional, Tuple
from urllib.parse import urlparse

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .types import SignatureError

_SIGNATURE_LABEL = "sig1"
_REQUIRED_COMPONENTS = {"@method", "@target-uri", "@authority", "date"}


def _normalize_headers(headers: dict) -> Dict[str, str]:
    return {str(k).lower(): str(v) for k, v in headers.items()}


def _content_digest(body: bytes) -> str:
    digest = hashlib.sha256(body).digest()
    return f"sha-256=:{base64.b64encode(digest).decode()}:"


def _format_signature_params(components: List[str], params: List[str]) -> str:
    components_str = " ".join(f'"{c}"' for c in components)
    if params:
        return f"({components_str});" + ";".join(params)
    return f"({components_str})"


def _normalize_authority(parsed) -> str:
    if not parsed.hostname:
        return parsed.netloc
    host = parsed.hostname.lower()
    scheme = (parsed.scheme or "").lower()
    port = parsed.port
    default_port = 80 if scheme == "http" else 443 if scheme == "https" else None

    if ":" in host:
        host_value = f"[{host}]"
    else:
        host_value = host

    if port and port != default_port:
        return f"{host_value}:{port}"
    return host_value


def _build_signature_base(
    method: str,
    url: str,
    headers: dict,
    components: List[str],
    signature_params: str,
) -> str:
    parsed = urlparse(url)
    header_lookup = _normalize_headers(headers)

    lines: List[str] = []
    for component in components:
        if component == "@method":
            value = method.upper()
        elif component == "@target-uri":
            value = url
        elif component == "@authority":
            value = _normalize_authority(parsed)
        elif component == "@scheme":
            value = (parsed.scheme or "").lower()
        elif component == "@path":
            value = parsed.path or "/"
        elif component == "@query":
            value = f"?{parsed.query}" if parsed.query else "?"
        elif component.startswith("@"):
            raise SignatureError(f"Unsupported signature component: {component}")
        else:
            lookup_key = component.lower()
            if lookup_key not in header_lookup:
                raise SignatureError(f"Missing required header for signature: {component}")
            value = header_lookup[lookup_key]
        lines.append(f'"{component}": {value}')

    lines.append(f'"@signature-params": {signature_params}')
    return "\n".join(lines)


def _parse_signature_header(signature_header: str) -> Tuple[str, bytes]:
    parts = [part.strip() for part in signature_header.split(",") if part.strip()]
    for part in parts:
        match = re.match(r'(?P<label>[a-zA-Z0-9_-]+)=:(?P<b64>[^:]+):', part)
        if not match:
            continue
        label = match.group("label")
        signature_b64 = match.group("b64")
        try:
            signature_bytes = base64.b64decode(signature_b64)
        except Exception:
            raise SignatureError("Invalid signature encoding")
        return label, signature_bytes
    raise SignatureError("Invalid Signature header format")


def _parse_signature_input(
    signature_input: str, label: str
) -> Tuple[List[str], List[str], int, Optional[str]]:
    parts = [part.strip() for part in signature_input.split(",") if part.strip()]
    target = None
    for part in parts:
        if part.startswith(f"{label}="):
            target = part
            break
    if not target:
        raise SignatureError("Signature-Input missing matching label")

    _, value = target.split("=", 1)
    if not value.startswith("("):
        raise SignatureError("Invalid Signature-Input header format")

    closing_index = value.find(")")
    if closing_index == -1:
        raise SignatureError("Invalid Signature-Input header format")

    components_str = value[1:closing_index].strip()
    params_str = value[closing_index + 1 :].strip()
    if params_str.startswith(";"):
        params_str = params_str[1:]

    components = re.findall(r'"([^"]+)"', components_str)
    if not components:
        raise SignatureError("Signature-Input missing components")

    if params_str:
        params = [p.strip() for p in params_str.split(";") if p.strip()]
    else:
        params = []
    params_map: Dict[str, str] = {}
    for param in params:
        if "=" in param:
            key, val = param.split("=", 1)
            params_map[key] = val
        else:
            params_map[param] = ""

    created_val = params_map.get("created")
    if created_val is None:
        raise SignatureError("Missing created timestamp")
    try:
        created = int(created_val)
    except ValueError:
        raise SignatureError("Invalid created timestamp")

    alg = params_map.get("alg")
    if alg:
        alg = alg.strip('"')

    return components, params, created, alg


def generate_keypair() -> Tuple[str, str]:
    """Generate a new Ed25519 keypair.

    Returns:
        Tuple of (private_key_b64, public_key_b64)
    """
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()

    private_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    public_bytes = public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )

    return (
        base64.b64encode(private_bytes).decode(),
        base64.b64encode(public_bytes).decode(),
    )


def load_private_key(private_key_b64: str) -> Ed25519PrivateKey:
    """Load an Ed25519 private key from base64."""
    private_bytes = base64.b64decode(private_key_b64)
    return Ed25519PrivateKey.from_private_bytes(private_bytes)


def load_public_key(public_key_b64: str) -> Ed25519PublicKey:
    """Load an Ed25519 public key from base64."""
    public_bytes = base64.b64decode(public_key_b64)
    return Ed25519PublicKey.from_public_bytes(public_bytes)


def create_signature_base(
    method: str,
    url: str,
    headers: dict,
    body: Optional[bytes] = None,
    created: Optional[int] = None,
    date: Optional[str] = None,
) -> Tuple[str, dict, str]:
    """Create the signature base string and required headers.

    Following RFC 9421 HTTP Message Signatures.

    Args:
        method: HTTP method (GET, POST, etc.)
        url: Full request URL
        headers: Existing headers dict (will be modified)
        body: Request body bytes (optional)
        created: Override created timestamp (Unix epoch)
        date: Override Date header if not already present

    Returns:
        Tuple of (signature_base_string, updated_headers, signature_params)
    """
    parsed = urlparse(url)

    # Ensure required headers exist
    now = datetime.now(timezone.utc)
    headers = dict(headers)  # Copy
    header_lookup = _normalize_headers(headers)

    if "host" not in header_lookup:
        headers["Host"] = parsed.netloc
        header_lookup["host"] = parsed.netloc

    if "date" not in header_lookup:
        if date:
            headers["Date"] = date
        else:
            headers["Date"] = now.strftime("%a, %d %b %Y %H:%M:%S GMT")
        header_lookup["date"] = headers["Date"]

    # Add created timestamp (Unix epoch)
    if created is None:
        created = int(now.timestamp())

    # Build the signature base (RFC 9421 format)
    components = ["@method", "@target-uri", "@authority", "date"]

    if body is not None and len(body) > 0:
        headers["Content-Digest"] = _content_digest(body)
        components.append("content-digest")

    signature_params = _format_signature_params(
        components, [f"created={created}", 'alg="ed25519"']
    )

    signature_base = _build_signature_base(method, url, headers, components, signature_params)
    return signature_base, headers, signature_params


def sign_request(
    method: str,
    url: str,
    headers: dict,
    body: Optional[bytes],
    private_key: Ed25519PrivateKey,
    key_id: str,
    created: Optional[int] = None,
    date: Optional[str] = None,
) -> dict:
    """Sign an HTTP request.

    Args:
        method: HTTP method
        url: Full request URL
        headers: Request headers
        body: Request body (optional)
        private_key: Ed25519 private key
        key_id: Key identifier (agent username)
        created: Override created timestamp (Unix epoch)
        date: Override Date header if not already present

    Returns:
        Updated headers dict with Signature and Signature-Input headers
    """
    signature_base, headers, signature_params = create_signature_base(
        method, url, headers, body, created=created, date=date
    )

    # Sign the base string
    signature_bytes = private_key.sign(signature_base.encode())
    signature_b64 = base64.b64encode(signature_bytes).decode()

    # Build the signature headers (RFC 9421)
    headers["Signature-Input"] = f"{_SIGNATURE_LABEL}={signature_params}"
    headers["Signature"] = f"{_SIGNATURE_LABEL}=:{signature_b64}:"

    # Add key ID for verification lookup
    headers["X-MoltAuth-Key-Id"] = key_id

    return headers


def verify_signature(
    method: str,
    url: str,
    headers: dict,
    body: Optional[bytes],
    public_key: Ed25519PublicKey,
    max_age_seconds: Optional[int] = 300,
    max_clock_skew_seconds: Optional[int] = 60,
    required_components: Optional[set] = _REQUIRED_COMPONENTS,
    require_content_digest: bool = True,
) -> bool:
    """Verify an HTTP request signature.

    Args:
        method: HTTP method
        url: Full request URL
        headers: Request headers (must include Signature and Signature-Input)
        body: Request body (optional)
        public_key: Ed25519 public key
        max_age_seconds: Maximum age of signature (default: 5 minutes, None disables)
        max_clock_skew_seconds: Allowed clock skew for created timestamp
        required_components: Components required to be present in signature
        require_content_digest: Require content-digest when body is present

    Returns:
        True if signature is valid

    Raises:
        SignatureError: If signature is invalid or expired
    """
    header_lookup = _normalize_headers(headers)

    sig_input = header_lookup.get("signature-input", "")
    sig_header = header_lookup.get("signature", "")

    if not sig_input or not sig_header:
        raise SignatureError("Missing Signature or Signature-Input header")

    label, signature_bytes = _parse_signature_header(sig_header)
    components, params, created, alg = _parse_signature_input(sig_input, label)

    if alg and alg != "ed25519":
        raise SignatureError(f"Unsupported signature algorithm: {alg}")

    # Check signature age
    now = int(time.time())
    if max_age_seconds is not None and now - created > max_age_seconds:
        raise SignatureError(
            f"Signature expired (age: {now - created}s, max: {max_age_seconds}s)"
        )

    if max_clock_skew_seconds is not None and created > now + max_clock_skew_seconds:
        raise SignatureError("Signature created in the future")

    if required_components:
        missing = required_components - set(components)
        if missing:
            missing_list = ", ".join(sorted(missing))
            raise SignatureError(f"Signature missing required components: {missing_list}")

    if require_content_digest and body is not None and len(body) > 0:
        if "content-digest" not in components:
            raise SignatureError("Missing content-digest signature component for request body")

    if "content-digest" in components:
        digest_header = header_lookup.get("content-digest")
        if not digest_header:
            raise SignatureError("Missing Content-Digest header")
        body_bytes = body if body is not None else b""
        expected_digest = _content_digest(body_bytes)
        if digest_header != expected_digest:
            raise SignatureError("Content-Digest mismatch")

    signature_params = _format_signature_params(components, params)
    signature_base = _build_signature_base(method, url, headers, components, signature_params)

    # Verify
    try:
        public_key.verify(signature_bytes, signature_base.encode())
        return True
    except Exception:
        raise SignatureError("Signature verification failed")


def extract_key_id(headers: dict) -> Optional[str]:
    """Extract the key ID (agent username) from request headers."""
    header_lookup = {k.lower(): v for k, v in headers.items()}
    return header_lookup.get("x-moltauth-key-id")

File: js/src/signing.ts
/**
 * HTTP Signatures implementation using Ed25519.
 *
 * Implements RFC 9421 (HTTP Message Signatures) with Ed25519.
 */

import { createHash, sign, verify, generateKeyPairSync, KeyObject } from 'crypto';
import { SignatureError } from './types';

const SIGNATURE_LABEL = 'sig1';
const REQUIRED_COMPONENTS = new Set(['@method', '@target-uri', '@authority', 'date']);

function normalizeHeaders(headers: Record<string, string>): Record<string, string> {
  const normalized: Record<string, string> = {};
  for (const [key, value] of Object.entries(headers)) {
    normalized[key.toLowerCase()] = String(value);
  }
  return normalized;
}

function normalizeAuthority(url: URL): string {
  const hostname = url.hostname.toLowerCase();
  const scheme = url.protocol.replace(':', '').toLowerCase();
  const port = url.port ? Number.parseInt(url.port, 10) : undefined;
  const defaultPort = scheme === 'http' ? 80 : scheme === 'https' ? 443 : undefined;
  const needsPort = port !== undefined && port !== defaultPort;

  const hostValue = hostname.includes(':') ? `[${hostname}]` : hostname;
  if (needsPort) {
    return `${hostValue}:${port}`;
  }
  return hostValue;
}

function contentDigest(body: Buffer): string {
  const digest = createHash('sha256').update(body).digest('base64');
  return `sha-256=:${digest}:`;
}

function formatSignatureParams(components: string[], params: string[]): string {
  const componentsStr = components.map((c) => `"${c}"`).join(' ');
  if (params.length > 0) {
    return `(${componentsStr});${params.join(';')}`;
  }
  return `(${componentsStr})`;
}

function buildSignatureBase(
  method: string,
  url: string,
  headers: Record<string, string>,
  components: string[],
  signatureParams: string
): string {
  const parsedUrl = new URL(url);
  const headerLookup = normalizeHeaders(headers);

  const lines: string[] = [];
  for (const component of components) {
    if (component === '@method') {
      lines.push(`"@method": ${method.toUpperCase()}`);
    } else if (component === '@target-uri') {
      lines.push(`"@target-uri": ${url}`);
    } else if (component === '@authority') {
      lines.push(`"@authority": ${normalizeAuthority(parsedUrl)}`);
    } else if (component === '@scheme') {
      lines.push(`"@scheme": ${parsedUrl.protocol.replace(':', '').toLowerCase()}`);
    } else if (component === '@path') {
      lines.push(`"@path": ${parsedUrl.pathname || '/'}`);
    } else if (component === '@query') {
      lines.push(`"@query": ${parsedUrl.search ? parsedUrl.search : '?'}`);
    } else if (component.startsWith('@')) {
      throw new SignatureError(`Unsupported signature component: ${component}`);
    } else {
      const lookupKey = component.toLowerCase();
      if (!headerLookup[lookupKey]) {
        throw new SignatureError(`Missing required header for signature: ${component}`);
      }
      lines.push(`"${component}": ${headerLookup[lookupKey]}`);
    }
  }

  lines.push(`"@signature-params": ${signatureParams}`);
  return lines.join('\n');
}

function parseSignatureHeader(signatureHeader: string): { label: string; signature: Buffer } {
  const parts = signatureHeader.split(',').map((part) => part.trim()).filter(Boolean);
  for (const part of parts) {
    const match = part.match(/^([a-zA-Z0-9_-]+)=:([^:]+):$/);
    if (!match) continue;
    const [, label, signatureBase64] = match;
    try {
      return { label, signature: Buffer.from(signatureBase64, 'base64') };
    } catch {
      throw new SignatureError('Invalid signature encoding');
    }
  }
  throw new SignatureError('Invalid Signature header format');
}

function parseSignatureInput(
  signatureInput: string,
  label: string
): { components: string[]; params: string[]; created: number; alg?: string } {
  const parts = signatureInput.split(',').map((part) => part.trim()).filter(Boolean);
  const target = parts.find((part) => part.startsWith(`${label}=`));
  if (!target) {
    throw new SignatureError('Signature-Input missing matching label');
  }

  const separatorIndex = target.indexOf('=');
  const value = separatorIndex >= 0 ? target.slice(separatorIndex + 1) : '';
  if (!value.startsWith('(')) {
    throw new SignatureError('Invalid Signature-Input header format');
  }
  const closingIndex = value.indexOf(')');
  if (closingIndex === -1) {
    throw new SignatureError('Invalid Signature-Input header format');
  }

  const componentsStr = value.slice(1, closingIndex).trim();
  let paramsStr = value.slice(closingIndex + 1).trim();
  if (paramsStr.startsWith(';')) {
    paramsStr = paramsStr.slice(1);
  }

  const components = Array.from(componentsStr.matchAll(/"([^"]+)"/g)).map((m) => m[1]);
  if (components.length === 0) {
    throw new SignatureError('Signature-Input missing components');
  }

  const params = paramsStr
    ? paramsStr.split(';').map((param) => param.trim()).filter(Boolean)
    : [];
  const paramsMap: Record<string, string> = {};
  for (const param of params) {
    const [key, valuePart] = param.split('=', 2);
    paramsMap[key] = valuePart ?? '';
  }

  if (!paramsMap.created) {
    throw new SignatureError('Missing created timestamp');
  }
  const created = Number.parseInt(paramsMap.created, 10);
  if (!Number.isFinite(created)) {
    throw new SignatureError('Invalid created timestamp');
  }

  const alg = paramsMap.alg ? paramsMap.alg.replace(/"/g, '') : undefined;
  return { components, params, created, alg };
}

/**
 * Generate a new Ed25519 keypair.
 *
 * @returns Tuple of [privateKeyBase64, publicKeyBase64]
 */
export function generateKeypair(): [string, string] {
  const { publicKey, privateKey } = generateKeyPairSync('ed25519');

  const privateBytes = privateKey.export({ type: 'pkcs8', format: 'der' });
  const publicBytes = publicKey.export({ type: 'spki', format: 'der' });

  // Extract raw key bytes (skip DER headers)
  // Ed25519 private key in PKCS8: 48 bytes total, raw key starts at offset 16
  // Ed25519 public key in SPKI: 44 bytes total, raw key starts at offset 12
  const privateRaw = privateBytes.subarray(16, 48);
  const publicRaw = publicBytes.subarray(12, 44);

  return [privateRaw.toString('base64'), publicRaw.toString('base64')];
}

/**
 * Load an Ed25519 private key from base64.
 */
export function loadPrivateKey(privateKeyBase64: string): KeyObject {
  const privateRaw = Buffer.from(privateKeyBase64, 'base64');

  // Wrap raw bytes in PKCS8 DER format
  const pkcs8Header = Buffer.from([
    0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70,
    0x04, 0x22, 0x04, 0x20,
  ]);
  const pkcs8Der = Buffer.concat([pkcs8Header, privateRaw]);

  return require('crypto').createPrivateKey({
    key: pkcs8Der,
    format: 'der',
    type: 'pkcs8',
  });
}

/**
 * Load an Ed25519 public key from base64.
 */
export function loadPublicKey(publicKeyBase64: string): KeyObject {
  const publicRaw = Buffer.from(publicKeyBase64, 'base64');

  // Wrap raw bytes in SPKI DER format
  const spkiHeader = Buffer.from([
    0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
  ]);
  const spkiDer = Buffer.concat([spkiHeader, publicRaw]);

  return require('crypto').createPublicKey({
    key: spkiDer,
    format: 'der',
    type: 'spki',
  });
}

interface SignatureComponents {
  signatureBase: string;
  signatureParams: string;
  headers: Record<string, string>;
}

/**
 * Create the signature base string and required headers.
 */
function createSignatureBase(
  method: string,
  url: string,
  headers: Record<string, string>,
  body?: Buffer,
  created?: number,
  date?: string
): SignatureComponents {
  const parsedUrl = new URL(url);
  const now = new Date();

  // Ensure required headers exist
  headers = { ...headers };
  const headerLookup = normalizeHeaders(headers);

  if (!headerLookup.host) {
    headers['Host'] = parsedUrl.host;
    headerLookup.host = parsedUrl.host;
  }

  if (!headerLookup.date) {
    headers['Date'] = date ?? now.toUTCString();
    headerLookup.date = headers['Date'];
  }

  const createdTimestamp = created ?? Math.floor(now.getTime() / 1000);

  const components = ['@method', '@target-uri', '@authority', 'date'];
  if (body && body.length > 0) {
    headers['Content-Digest'] = contentDigest(body);
    components.push('content-digest');
  }

  const signatureParams = formatSignatureParams(components, [
    `created=${createdTimestamp}`,
    'alg="ed25519"',
  ]);
  const signatureBase = buildSignatureBase(method, url, headers, components, signatureParams);

  return { signatureBase, signatureParams, headers };
}

/**
 * Sign an HTTP request.
 */
export function signRequest(
  method: string,
  url: string,
  headers: Record<string, string>,
  body: Buffer | undefined,
  privateKey: KeyObject,
  keyId: string,
  created?: number,
  date?: string
): Record<string, string> {
  const { signatureBase, signatureParams, headers: updatedHeaders } = createSignatureBase(
    method,
    url,
    headers,
    body,
    created,
    date
  );

  // Sign the base string
  const signature = sign(null, Buffer.from(signatureBase), privateKey);
  const signatureBase64 = signature.toString('base64');

  // Build signature headers (RFC 9421)
  updatedHeaders['Signature-Input'] = `${SIGNATURE_LABEL}=${signatureParams}`;
  updatedHeaders['Signature'] = `${SIGNATURE_LABEL}=:${signatureBase64}:`;
  updatedHeaders['X-MoltAuth-Key-Id'] = keyId;

  return updatedHeaders;
}

/**
 * Verify an HTTP request signature.
 */
export function verifySignature(
  method: string,
  url: string,
  headers: Record<string, string>,
  body: Buffer | undefined,
  publicKey: KeyObject,
  maxAgeSeconds: number | null = 300,
  maxClockSkewSeconds: number | null = 60,
  requiredComponents: string[] = Array.from(REQUIRED_COMPONENTS),
  requireContentDigest: boolean = true
): boolean {
  const headerLookup = normalizeHeaders(headers);

  const sigInput = headerLookup['signature-input'] || '';
  const sigHeader = headerLookup['signature'] || '';

  if (!sigInput || !sigHeader) {
    throw new SignatureError('Missing Signature or Signature-Input header');
  }

  const { label, signature } = parseSignatureHeader(sigHeader);
  const { components, params, created, alg } = parseSignatureInput(sigInput, label);

  if (alg && alg !== 'ed25519') {
    throw new SignatureError(`Unsupported signature algorithm: ${alg}`);
  }

  const now = Math.floor(Date.now() / 1000);

  // Check signature age
  if (maxAgeSeconds !== null && maxAgeSeconds !== undefined && now - created > maxAgeSeconds) {
    throw new SignatureError(
      `Signature expired (age: ${now - created}s, max: ${maxAgeSeconds}s)`
    );
  }

  if (
    maxClockSkewSeconds !== null &&
    maxClockSkewSeconds !== undefined &&
    created > now + maxClockSkewSeconds
  ) {
    throw new SignatureError('Signature created in the future');
  }

  if (requiredComponents && requiredComponents.length > 0) {
    const missing = requiredComponents.filter((c) => !components.includes(c));
    if (missing.length > 0) {
      throw new SignatureError(`Signature missing required components: ${missing.join(', ')}`);
    }
  }

  if (requireContentDigest && body && body.length > 0 && !components.includes('content-digest')) {
    throw new SignatureError('Missing content-digest signature component for request body');
  }

  if (components.includes('content-digest')) {
    const digestHeader = headerLookup['content-digest'];
    if (!digestHeader) {
      throw new SignatureError('Missing Content-Digest header');
    }
    const bodyBytes = body ?? Buffer.alloc(0);
    const expectedDigest = contentDigest(bodyBytes);
    if (digestHeader !== expectedDigest) {
      throw new SignatureError('Content-Digest mismatch');
    }
  }

  const signatureParams = formatSignatureParams(components, params);
  const signatureBase = buildSignatureBase(method, url, headers, components, signatureParams);

  // Verify
  try {
    const isValid = verify(null, Buffer.from(signatureBase), publicKey, signature);
    if (!isValid) {
      throw new SignatureError('Signature verification failed');
    }
    return true;
  } catch (e) {
    if (e instanceof SignatureError) throw e;
    throw new SignatureError('Signature verification failed');
  }
}

/**
 * Extract the key ID (agent username) from request headers.
 */
export function extractKeyId(headers: Record<string, string>): string | undefined {
  const headerLookup: Record<string, string> = {};
  for (const [key, value] of Object.entries(headers)) {
    headerLookup[key.toLowerCase()] = value;
  }
  return headerLookup['x-moltauth-key-id'];
}
